sdks/python/apache_beam/transforms/external.py

```diff
@@ -42,6 +44,63 @@
   grpc = None
 # pylint: enable=wrong-import-order, wrong-import-position, ungrouped-imports
 
+DEFAULT_EXPANSION_SERVICE = 'localhost:8097'
+FILTERED_CODERS = {
+  'beam:coder:length_prefix:v1',
+}
+
+
+def iter_urns(coder, context=None):
+  yield coder.to_runner_api_parameter(context)[0]
+  for child in coder._get_component_coders():
+    for urn in iter_urns(child, context):
+      yield urn
+
+
+class External(ptransform.PTransform):
+  """
+  Base class for transforms that produce external transforms.
+  """
+  _urn = None
+
+  def __init__(self, expansion_service=None):
+    self.expansion_service = expansion_service or DEFAULT_EXPANSION_SERVICE
+
+  def get_config_args(self):
+    """
+    :return: dictionary of encoded configuration values to pass to
+             ExternalConfigurationPayload
+    """
+    raise NotImplementedError
+
+  @classmethod
+  def config_value(cls, obj, coder):
+    """
+    Helper to create a ConfigValue with an encoded value.
+    """
+    return ConfigValue(
+      coder_urn=[urn for urn in iter_urns(coder)
+                 if urn not in FILTERED_CODERS],
```

Comment:
I don't think we can safely filter out certain coders, as that changes the encoding.

Reply:
Glad you mentioned it :)

I'm confused about why the `LengthPrefixCoder` urn is not included in the `ConfigValue .coder_urns` list in the original code.  The original list or the kafka `consumer_config` param was `['beam:coder:iterable:v1', 'beam:coder:kv:v1', 'beam:coder:bytes:v1', 'beam:coder:bytes:v1']`.  I was trying to keep the end result the same, and I assumed that there were coders which were "write-only", and thus did not need to be transmitted to Java.

Reply:
We do not need the length prefix coder here, do we? That's because we get the encoded bytes via Protobuf and do not have to reason about the length.

Reply:
But we do need the lengths if it is, say, KV<LengthPrefixedX, LenthPrefixedY>. What's more, we're getting the already encoded value here (which is different than if we had tried to filter them out before doing the encoding. 

Are you seeing instances of LengthPrefixCoder that need filtering out? If so, this deserves further investigation. Otherwise, let's just drop this.

Reply:
I removed this filter.  

Note that between this change and the `StrUtf8Coder` change, the list of coder urns produced by `ReadFromKafka` has changed from this:

```
 coder_urn: "beam:coder:iterable:v1"
 coder_urn: "beam:coder:kv:v1"
 coder_urn: "beam:coder:bytes:v1"
 coder_urn: "beam:coder:bytes:v1"
```

to this:

```
 coder_urn: "beam:coder:iterable:v1"
 coder_urn: "beam:coder:kv:v1"
 coder_urn: "beam:coder:length_prefix:v1"
 coder_urn: "beam:coder:string_utf8:v1"
 coder_urn: "beam:coder:length_prefix:v1"
 coder_urn: "beam:coder:string_utf8:v1"
```

Likewise for `WriteToKafka`,

@mxm  can you please confirm whether this will require adjustments to the Kafka Java code?

Reply:
Point taken about the length prefix being required. 

>@mxm can you please confirm whether this will require adjustments to the Kafka Java code?

For the length prefix to be supported, we will add a detection here to properly initialize the coder's component: https://github.com/apache/beam/blob/4e40b53198870b2af1b178bda8ffa326a60833dc/runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/expansion/ExpansionService.java#L210 Like already mentioned the encoding scheme is flawed and should very soon be replaced by the Components structure which would hold the element coder instead.

Reply:
> For the length prefix to be supported, we will add a detection here to properly initialize the coder's component:

There's one more wrinkle here:  now that the design has been changed to automatically detect the coder from the type hint, the `LengthPrefixCoder` is not being used at all.

e.g., given this type hint:

```
Iterable[Tuple[unicode, unicode]]
```

We generate this list of coders:

```
 coder_urn: "beam:coder:iterable:v1"
 coder_urn: "beam:coder:kv:v1"
 coder_urn: "beam:coder:string_utf8:v1"
 coder_urn: "beam:coder:string_utf8:v1"
```

How important is the `LengthPrefixCoder` in this case?  Can we drop it?  

@robertwb  If `LengthPrefixCoder` is required, do you have any thoughts on an interface for inserting this automatically?  i.e. Is there anything about a typehint that could indicate when it should automatically be used, or is it always an explicit choice by the user?

Reply:
My question about the `LengthPrefixCoder` is still outstanding, but I have one more for @mxm:  

I realized that I don't understand how the name mapping works for the payloads: e.g. how does  `"elements_per_period": 20` in our python-based payload end up calling `GenerateSequence.Builder.setElementsPerPeriod(20)` in Java?  I don't see any name mapping happening in Java, I don't think it can be based on order (dicts in python are not ordered, plus we omit optional values), so I'm a bit confused.

Reply:
We should not worry about LengthPrefixCoder at all. It's mostly for actors that need to pass records through without interpreting them. Eventually this can all be replaced by schemas anyway, so not something to worry to much about here.

Reply:
> We should not worry about LengthPrefixCoder at all.

Sounds good.

> Eventually this can all be replaced by schemas anyway, so not something to worry to much about here.

Can you point me at a document or issue for this?  I'd love to learn more.

Reply:
I played around with this some more.  The current design seems to *require* the LengthPrefixCoder.  If I don't wrap the str/bytes coder with length-prefix I get this error:

```
Caused by: org.apache.beam.sdk.coders.CoderException: java.io.EOFException: reached end of stream after reading 38 bytes; 112 bytes expected
```

@mxm  can you confirm that the current design requires the LengthPrefixCoder for strings?

@robertwb I saw the schema / row-coder PR at #9188.  Is this the schema support you were referring to?  Is anyone assigned to porting external transforms to using schema coders?  Luckily I think most of the discussion here about the design of the interface remains valid even after that's complete.

Reply:
Coders also have the notion of nested/unnested. It happens that LengthPrefix(UnnestedBytes) == NestedBytes. Perhaps this is part of the issue here?

Yes, that's the schema/row stuff. I was just commenting that this particular part is going to change (no, no one's on it yet) so whatever works here is fine.

Reply:
@chadrik Sorry for the late reply. Vacations and open-source work do not always go together well :) Actually, for the StringUtf8Coder, the length should be added automatically in the nested context, no need to add the length prefix.

>how does "elements_per_period": 20 in our python-based payload end up calling GenerateSequence.Builder.setElementsPerPeriod(20) in Java?

It is a simple mapping scheme which we also use for the pipeline options. It converts `snake_case` to `camelCase` and then looks up the setter in the Java configuration class.

>@mxm can you confirm that the current design requires the LengthPrefixCoder for strings?

I don't think so, even with the byte encoding in master, the length prefix is added automatically by the `ByteArrayCoder`.

Reply:
Perhaps paste the full stack trace here if you are still seeing problems.